onnxruntime/contrib_ops/cuda/bert/cutlass_fmha/memory_efficient_attention.cu

```diff
@@ -0,0 +1,152 @@
+
+    p.causal = params.causal;
+  }
+
+  constexpr auto kernel_fn = attention_kernel_batched_impl<Attention>;
+  int smem_bytes = sizeof(typename Attention::SharedStorage);
+  if (smem_bytes > 0xc000) {
+    ORT_ENFORCE(params.sm >= 70, "This kernel requires too much shared memory on this machine!");
+    static bool once = [&]() {
+      cudaFuncSetAttribute(kernel_fn, cudaFuncAttributeMaxDynamicSharedMemorySize, smem_bytes);
+      return true;
+    }();
+  }
+
+  ORT_ENFORCE(Attention::check_supported(p));
+  kernel_fn<<<p.getBlocksGrid(), p.getThreadsGrid(), smem_bytes, params.stream>>>(p);
+}
+
+template <typename T, typename ArchTag, int queries_per_block, int keys_per_block, bool single_value_iteration>
+void DispatchIsAligned(const MemoryEfficientAttentionParams& params) {
+  using AlignedAK = AttentionKernel<T, ArchTag, true, queries_per_block, keys_per_block, single_value_iteration>;
+
+  // Run a more efficient kernel with `isAligned=True` when memory is correctly aligned.
+  bool is_aligned = params.qk_head_size % AlignedAK::kAlignmentQ == 0 &&
+                    params.qk_head_size % AlignedAK::kAlignmentK == 0 &&
+                    params.v_head_size % AlignedAK::kAlignmentV == 0;
+
+  DISPATCH_BOOL(is_aligned, kIsAligned, ([&]() {
+                  LaunchCutlassFmha<T, ArchTag, kIsAligned, queries_per_block, keys_per_block, single_value_iteration>(params);
+                }));
+}
+
+template <typename T, typename ArchTag>
+void DispatchBlockSize(const MemoryEfficientAttentionParams& params) {
+  if (params.v_head_size <= 64) {
+    DispatchIsAligned<T, ArchTag, 64, 64, true>(params);
+  } else if (params.v_head_size <= 128) {
+    DispatchIsAligned<T, ArchTag, 32, 128, true>(params);
+  } else {
+    DispatchIsAligned<T, ArchTag, 32, 128, false>(params);
+  }
+}
+
+template <typename T>
+void DispatchArchTag(const MemoryEfficientAttentionParams& params) {
+  const int32_t& sm = params.sm;
+  if (sm >= 80) {
+    DispatchBlockSize<T, cutlass::arch::Sm80>(params);
+  } else if (sm >= 75) {
+    DispatchBlockSize<T, cutlass::arch::Sm75>(params);
+  } else if (sm >= 70) {
+    DispatchBlockSize<T, cutlass::arch::Sm70>(params);
+  } else if (sm >= 50) {
+    DispatchBlockSize<T, cutlass::arch::Sm50>(params);
+  } else {
+  }
+}
+
+common::Status run_memory_efficient_attention(const MemoryEfficientAttentionParams& params) {
+  if (params.is_half) {
+    DispatchArchTag<cutlass::half_t>(params);
+  } else {
+    DispatchArchTag<float>(params);
+  }
+
+  CUDA_RETURN_IF_ERROR(cudaStreamSynchronize(params.stream));
+  return Status::OK();
+}
+
+bool has_memory_efficient_attention(int32_t sm, bool is_half) {
+  return sm >= (is_half ? 53 : 50);
+}
+
+}  // namespace cuda
+}  // namespace contrib
+}  // namespace onnxruntime
+
+#if defined(__GNUC__)
+#pragma GCC diagnostic pop
+#endif
+
+#else  // NOT USE_FLASH_ATTENTION
+
+bool has_memory_efficient_attention(int32_t, bool) {
```

Comment:
>has_memory_efficient_attention [](http://example.com/codeflow?start=5&length=30)

It should also be under ns onnxruntime::contrib::cuda

Reply:
good catch